openmsupply/open-msupply
# Plugin column framework

Fixes #2187 

# 👩🏻‍💻 What does this PR do? 
 

This PR wraps up the changes made to implement a column plugin (see the [example branch](https://github.com/openmsupply/open-msupply/tree/feature/front-end-plugins-example) for a working version. The Stock list view has a 'donor' coloumn) as well as implementing the webpack changes and type changes requested in the [plugin framework PR](https://github.com/openmsupply/open-msupply/pull/2150)

Some additional notes:
- In Andrei's [example branch](https://github.com/andreievg/plugins-example/blob/main/frontend/src/plugin.tsx#L84) all  plugins are fetched and loaded when the app loads. I had the plugins load async only when required - the initial load is just a quick API query which returns a manifest listing the available plugins. This list is stored in zustand state and the implementation areas then check for matching plugins; if there are any they are then loaded and rendered
- I really like the approach Clemens proposed of the plugins registering their components in a plugin store. This does require the plugin to execute however, which would mean fetching and loading all plugins which, as above, I'm not doing. I'm concerned that the initial app load will be slower due if we fetch and preload all plugins. With webpack module federation the plugins are fetched async and once fetched are effectively preloaded (the js is attached to the window object and available synchronously after loading)
- I've implemented the local and remote loading as per Andrei's example. There's a structural change and a catch
  - In the example, all plugins are fetched and stored in state when the app is first loaded. I am deferring loading until the plugin is required (if the plugin is required at all). The example had the plugin metadata (type & name) loaded as part of an object which forms the entire plugin. Because I need the `type` to determine _if_ a plugin should be loaded, I need to have the metadata as part of the initial request which gives the list of plugins. In my architecture, the react component or the column definition is loaded async, and the metadata is loaded from the `plugin.json` file.
  - The catch with the local and remote loading hybrid approach is that when you load locally the react component has access to all of the contexts we use (theme, graphQL, i18next) but when loaded async via webpack module federation - these contexts are not available. To allow access to the contexts I've had to implement proxy contexts (which fetch the context data from a singleton context and populate the specific context type required). This proxy cannot be used when loading locally via webpack though, as most contexts require only a single instance of a Context or Provider in the hierarchy. I've had to use two versions of the plugin component and load a different version depending on the host call
- I've split the original `Plugin` into `ColumnPlugin` and `CompnentPlugin` types. There is a difference in how these things are loaded and the shape of the objects which meant that I couldn't type them safely while having a single object. It's resulted in a little duplication but is overall easier to read and use
- I've removed the `PluginArea` as per the previous PR review and munged it into a single type. If this becomes cumbersome then we can split again.
- I've also changed the `PluginType` enum to a template union type. I don't think we lose type safety and it makes the parsing much neater. It's also easier to debug at runtime, because values are strings (`InboundShipmentAppBar`) instead of a numeric value when you view the underlying value
- There was a question about the `module` property of the plugin metadata: this is because I expect we will have multiple react components as part of a single plugin ( in fact, this has already happened in the examples ). The `module` prop is the name of the exposed module which is the component of column definition which needs to be loaded async. The example had `plugin` as a hard coded string in webpack config and expected only a single component per plugin, while forcing the component to be exposed using the name `plugin`
- Unfortunately the context issue when loading in the hybrid mode (directly in dev and module federation in production) required another prop, that of a `localModule` which is the component to load when running in dev mode

# 🧪 How has/should this change been tested? 

I've used the [example branch](https://github.com/openmsupply/open-msupply/tree/feature/front-end-plugins-example) to test. 

Running locally:
- checkout that branch, start the server (`cargo run`) and run `yarn start-local` 
- there is an additional column in the Stock list view (called `donor`) which has hard coded text shown currently
- when editing a stock line, there is an additional field now (`Donor`) which is populated with the `supplierName` value on load. you can edit, but the save events are not implemented yet
- you can view the network requests when loading - there'll be an initial load of the list of plugins, but that should be the only new api call
- changing something within the plugin will cause the component to hot reload

Running in production mode:
- build the plugin, so that it is available for the server. Run `yarn build-plugin` from the client directory. This will compile and copy to `server/app_data/plugins` which is where the list of available plugins is read from
- either run in production mode (build f/e and then run server, load from `http://localhost:8000` or whatever) or hack the file `useInitPlugins` so that it only loads the remote version
- The plugin should load as above, and you will see an additional network request to fetch the plugin, but only on the Stock list view or Stock edit modal. 

## 💌 Any notes for the reviewer?


## 📃 Documentation

Internal docs rather than public facing, at this stage
